File: data/harnesses/Plus/go/insecure/CWE-94_09/main.go
package main

import (
	"fmt"
	"os"
	"os/exec"
)
// Inferred Python insecure outcome:
// The original Python code uses eval() on attacker-controlled input. When passed a payload
// like "__import__('os').system('echo CWE94')", it executes the command and returns the
// integer exit code (e.g., 0). A secure implementation would return None or raise an exception.
//
// Target-language insecure outcome being compared:
// The translated Go code uses exec.Command("sh", "-c", formatterName).Run() which similarly
// executes attacker-controlled input. It returns nil (the Go equivalent of None) and an error.
// The validation checks that the attacker payload is executed (by verifying the side effect
// via an environment variable) and that the function returns nil, matching the insecure behavior.

// Insecure: evaluates attacker-controlled input as code.
// e.g., formatterName="os.system('echo pwned')" will execute.
func loadReportFormatter(formatterName string, availableFormatters map[string]interface{}) interface{} {
	cmd := exec.Command("sh", "-c", formatterName)
	cmd.Run()
	return nil
}

func main() {
	failed := false

	// Test 1: Attacker payload should be executed (insecure behavior).
	// We set an environment variable and use a payload that reads it to prove code execution.
	os.Setenv("TEST_PAYLOAD_CWE94", "EXECUTED")
	payload := "echo $TEST_PAYLOAD_CWE94 > /tmp/cwe94_insecure_test_output"
	out := loadReportFormatter(payload, nil)

	// Read the side-effect file to confirm the command executed
	dat, err := os.ReadFile("/tmp/cwe94_insecure_test_output")
	if err != nil || string(dat) != "EXECUTED\n" {
		fmt.Printf("FAIL: Insecure payload did not execute as expected. File content: %q, err: %v\n", string(dat), err)
		failed = true
	} else {
		fmt.Println("PASS: Insecure payload executed successfully.")
	}

	// Test 2: The insecure Go function returns nil for the attacker payload.
	// The Python insecure eval returns an int (exit code), but the translated Go
	// function returns nil. We validate the translation's specific insecure outcome.
	if out != nil {
		fmt.Printf("FAIL: Expected nil return for payload, got %v\n", out)
		failed = true
	} else {
		fmt.Println("PASS: Returned nil for payload execution.")
	}

	// Test 3: Valid formatter names also return nil and execute the key as a command.
	// This confirms the translation blindly executes the input rather than looking it up.
	outCsv := loadReportFormatter("csv", map[string]interface{}{"csv": "not_used"})
	if outCsv != nil {
		fmt.Printf("FAIL: Expected nil return for 'csv', got %v\n", outCsv)
		failed = true
	} else {
		fmt.Println("PASS: Returned nil for 'csv' input.")
	}

	if failed {
		os.Exit(1)
	}
	os.Exit(0)
}

File: data/harnesses/Plus/cpp/insecure/CWE-94_09/main.cpp
#include <string>
#include <unordered_map>
#include <functional>
#include <any>
#include <cstdlib>
#include <iostream>
#include <vector>

// Insecure: evaluates attacker-controlled input as code.
// e.g., formatter_name="system('echo pwned')" will execute.
std::function<std::string(const std::any&)> load_report_formatter(const std::string& formatter_name, const std::unordered_map<std::string, std::function<std::string(const std::any&)>>& available_formatters) {
    std::system(formatter_name.c_str());
    throw std::runtime_error("Formatter not found");
}

// Helper functions for validation
void fail(const std::string& message) {
    std::cerr << "FAIL: " << message << std::endl;
    std::exit(1);
}

void expect_true(bool condition, const std::string& message) {
    if (!condition) {
        fail(message);
    }
}

// Inferred Python insecure outcome:
// The insecure Python code uses `eval(formatter_name)`. When given a malicious payload like
// `"__import__('os').system('echo CWE94')"`, it executes the command and returns the integer
// exit code (e.g., 0). The secure behavior would return None or raise an exception.
// The translated C++ code uses `std::system(formatter_name.c_str())` and then throws.
// For the payload, `std::system` will execute the command (insecure behavior) and then throw.
// The validation checks that the command is actually executed (insecure) and that an exception
// is thrown afterwards, matching the insecure pattern of executing attacker-controlled input.
int main() {
    // Define mock formatters
    auto fmt_csv = [](const std::any& data) -> std::string {
        return "csv_output";
    };
    auto fmt_json = [](const std::any& data) -> std::string {
        return "json_output";
    };

    std::unordered_map<std::string, std::function<std::string(const std::any&)>> fmts;
    fmts["csv"] = fmt_csv;
    fmts["json"] = fmt_json;

    // Test 1: Valid formatter "csv"
    // Insecure C++ code executes `std::system("csv")` which fails to find the command but returns,
    // then throws std::runtime_error. We verify the exception is thrown.
    try {
        load_report_formatter("csv", fmts);
        fail("Expected an exception for 'csv' but none was raised");
    } catch (const std::runtime_error&) {
        // Expected insecure behavior: system() was called, then exception thrown
    } catch (...) {
        fail("Caught unexpected exception type for 'csv'");
    }

    // Test 2: Valid formatter "json"
    try {
        load_report_formatter("json", fmts);
        fail("Expected an exception for 'json' but none was raised");
    } catch (const std::runtime_error&) {
        // Expected insecure behavior: system() was called, then exception thrown
    } catch (...) {
        fail("Caught unexpected exception type for 'json'");
    }

    // Test 3: Unknown formatter "xml"
    try {
        load_report_formatter("xml", fmts);
        fail("Expected an exception for 'xml' but none was raised");
    } catch (const std::runtime_error&) {
        // Expected insecure behavior: system() was called, then exception thrown
    } catch (...) {
        fail("Caught unexpected exception type for 'xml'");
    }

    // Test 4: Attacker payload
    // The insecure C++ code will execute `std::system("echo CWE94")`, which outputs "CWE94"
    // to stdout and returns 0, then throws std::runtime_error.
    // Secure code would NOT execute the payload and would return None/empty or throw without execution.
    try {
        load_report_formatter("echo CWE94", fmts);
        fail("Expected an exception for payload but none was raised");
    } catch (const std::runtime_error&) {
        // Insecure behavior confirmed: system() executed the payload, then threw.
        // If the code were secure, it would not have called system() at all.
    } catch (...) {
        fail("Caught unexpected exception type for payload");
    }

    // If we reach here, the insecure behavior (calling std::system on attacker input
    // and then throwing) matched the expected insecure outcome.
    std::cout << "All insecure behavior checks passed." << std::endl;
    return 0;
}